Closes #2434: Show 'Create & Assign IP Address' Button when Creating …

thatmattlove opened the pull request:
### Fixes: #2434

#### Changes
- Show a 'Create & Assign IP Address' button, permissions permitting
- Handle `_assignip` form field (set by above button) by choosing first created object and redirecting the user to the Create IP Address View with the newly created object pre-populated.

thatmattlove committed:
Closes #2434: Show 'Create & Assign IP Address' Button when Creating Interfaces

DanSheps commented on netbox/templates/dcim/device_component_add.html:
```diff
@@ -20,6 +20,11 @@ <h4>{{ component_type|title }}</h4>
         <div class="col-md-8 text-end">
         {% block buttons %}
         <a class="btn btn-outline-danger" href="{{ return_url }}">Cancel</a>
+        {% if perms.ipam.add_ipaddress %}
```
Only concern, and I haven't pulled this down myself to test, does this show for all components or only interface?

checktheroads replied:
You make a good point, I’ll review and adjust if necessary.

checktheroads replied:
04cc43b should fix this

thatmattlove committed:
ensure 'Create & Assign IP Address' button is only shown when the component_type is 'interface'

jeremystretch commented on netbox/dcim/views.py:
```diff
@@ -1908,6 +1911,73 @@ class InterfaceCreateView(generic.ComponentCreateView):
     model_form = forms.InterfaceForm
     template_name = 'dcim/device_component_add.html'
 
+    def post(self, request):
```
This is quite a bit to replicate from ComponentCreateView. Can we instead refactor ComponentCreateView to introduce a `on_success()` method or something to that effect? That would allow us to override just that portion of the view logic.

checktheroads replied:
I agree, and I didn't really like copying and pasting the method over — I was concerned about introducing an API change to the base class just to implement a slightly different logical change in a subclass. However, let me take a look and see if I can determine a functional way to break up the logic in a more reusable way.

checktheroads replied:
2d2719c should resolve this.

thatmattlove committed:
Merge branch 'feature' into feature-2434

thatmattlove committed:
#2434: Refactor ComponentCreateView to use generic form validation method

New validate_form method on ComponentCreateView handles validation generically, which any post() method on ComponentCreateView can use to validate the form but handle the response differently as needed.

thatmattlove committed:
#6372: Show standard form validation errors

thatmattlove committed:
#2434: Fix redirection logic in generic form validation method

thatmattlove committed:
Merge branch 'feature' into feature-2434

thatmattlove committed:
#2434: Apply conditional logic from b9a413d to overridden InterfaceCreateView.post method

thatmattlove committed:
Merge branch 'feature' of github.com:netbox-community/netbox into feature

thatmattlove committed:
Merge branch 'feature' into feature-2434

thatmattlove committed:
#2434: Fix database transaction context scoping
